src/api/scan/base_controller.go

```diff
@@ -110,118 +117,156 @@ func NewController() Controller {
 	}
 }
 
+func (bc *basicController) collectScanningArtifacts(ctx context.Context, r *scanner.Registration, artifact *ar.Artifact) ([]*ar.Artifact, bool, error) {
+	var (
+		scannable bool
+		artifacts []*ar.Artifact
+	)
+
+	walkFn := func(a *ar.Artifact) error {
+		hasCapability := HasCapability(r, a)
+
+		if !hasCapability && a.HasChildren() {
+			// image index not supported by the scanner, so continue to walk its children
+			return nil
+		}
+
+		artifacts = append(artifacts, a)
+
+		if hasCapability {
+			scannable = true
+			return ar.ErrSkip // this artifact supported by the scanner, skip to walk its children
+		}
+
+		return nil
+	}
+
+	if err := bc.ar.Walk(ctx, artifact, walkFn, nil); err != nil {
+		return nil, false, err
+	}
+
+	return artifacts, scannable, nil
+}
+
 // Scan ...
-func (bc *basicController) Scan(artifact *v1.Artifact, options ...Option) error {
+func (bc *basicController) Scan(ctx context.Context, artifact *ar.Artifact, options ...Option) error {
 	if artifact == nil {
 		return errors.New("nil artifact to scan")
 	}
 
-	// Parse options
-	ops, err := parseOptions(options...)
-	if err != nil {
-		return errors.Wrap(err, "scan controller: scan")
-	}
-
-	r, err := bc.sc.GetRegistrationByProject(artifact.NamespaceID)
+	r, err := bc.sc.GetRegistrationByProject(artifact.ProjectID)
 	if err != nil {
 		return errors.Wrap(err, "scan controller: scan")
 	}
 
 	// In case it does not exist
 	if r == nil {
-		return errs.WithCode(errs.PreconditionFailed, errs.Errorf("no available scanner for project: %d", artifact.NamespaceID))
+		return errs.WithCode(errs.PreconditionFailed, errs.Errorf("no available scanner for project: %d", artifact.ProjectID))
 	}
 
 	// Check if it is disabled
 	if r.Disabled {
 		return errs.WithCode(errs.PreconditionFailed, errs.Errorf("scanner %s is disabled", r.Name))
 	}
 
-	// Check the health of the registration by ping.
-	// The metadata of the scanner adapter is also returned.
-	meta, err := bc.sc.Ping(r)
+	artifacts, scannable, err := bc.collectScanningArtifacts(ctx, r, artifact)
 	if err != nil {
-		return errors.Wrap(err, "scan controller: scan")
+		return err
 	}
 
-	// Generate a UUID as track ID which groups the report records generated
-	// by the specified registration for the digest with given mime type.
-	trackID, err := bc.uuid()
-	if err != nil {
-		return errors.Wrap(err, "scan controller: scan")
+	if !scannable {
+		return errors.Errorf("the configured scanner %s does not support scanning artifact with mime type %s", r.Name, artifact.ManifestMediaType)
 	}
 
-	producesMimes := make([]string, 0)
-	matched := false
-	statusConflict := false
-	for _, ca := range meta.Capabilities {
-		for _, cm := range ca.ConsumesMimeTypes {
-			if cm == artifact.MimeType {
-				matched = true
-				break
-			}
-		}
+	type Param struct {
```

Comment:
what does `Param` mean?

Reply:
No meanings, just a temp values to hold the track id and produces mimes types which will be used to submit scan job later.